### DISCUSSION TITLE
How to integrate pyecharts with JScode in tooltip?

### DISCUSSION THREAD

**Discussion by TyrionZK** [zauberzeug/nicegui#2294] [Q&A]
### Description

How to integrate pyecharts with JScode in tooltip?

**Comment by falkoschindler** [comment]
Please be more specific.

To quote from our [FAQ](https://github.com/zauberzeug/nicegui/wiki/FAQs):
> When starting new questions or discussions, please put effort in formulating your post. Quite a number of people will read and think about your message. Make it worth their time.

**Comment by TyrionZK** [comment]
My problem is : 
I find nicegui can't render jscode in tooltip of pyecharts.

Problem pictures:
first picture pyecharts.png render rightly , but second picture nicegui.png render wrong.
![pyecharts](https://github.com/zauberzeug/nicegui/assets/23470438/2f8bf021-345e-42a9-a2d6-81399c787b79)
![nicegui](https://github.com/zauberzeug/nicegui/assets/23470438/e2a691a7-1ae3-4c98-9027-cdf721be5807)

My code is below:

```py
import json
from pyecharts import options as opts
from pyecharts.charts import Kline
from pyecharts.commons.utils import JsCode

from nicegui import ui

price = [
    [2320.26, 2320.26, 2287.3, 2362.94],
    [2300, 2291.3, 2288.26, 2308.38],
    [2295.35, 2346.5, 2295.35, 2345.92],
    [2347.22, 2358.98, 2337.35, 2363.8],
    [2360.75, 2382.48, 2347.89, 2383.76],
    [2383.43, 2385.42, 2371.23, 2391.82],
    [2377.41, 2419.02, 2369.57, 2421.15],
    [2425.92, 2428.15, 2417.58, 2440.38],
    [2411, 2433.13, 2403.3, 2437.42],
    [2432.68, 2334.48, 2427.7, 2441.73],
]

js_code_str = '''
    function(params){
        var name2index = new Array();
        for(var i=0;i<params.length;i++)
        {
            name2index[params[i].seriesName]=i;
        }
        var tooltip_data = [
            'trade_date: '+params[name2index['price']].name,
            'open: '+params[name2index['price']].value[1].toFixed(2),
            'close: '+params[name2index['price']].value[2].toFixed(2),
            'low: '+params[name2index['price']].value[3].toFixed(2),
            'high: '+params[name2index['price']].value[4].toFixed(2),
        ];
        var ret = tooltip_data.join('<br/>');
        return ret
    }
'''

kline = Kline()
kline.add_xaxis(["2017/7/{}".format(i + 1) for i in range(10)])
kline.add_yaxis(y_axis=price,series_name='price')
kline.set_global_opts(
    tooltip_opts=opts.TooltipOpts(
        trigger="axis",
        axis_pointer_type="cross",
        formatter=JsCode(js_code_str)
    ),
)

kline.render('pyecharts.html')

ui.echart(json.loads(kline.dump_options_with_quotes()))

ui.run()
```

**Comment by falkoschindler** [comment]
Ah, I see! Thanks for the clarification!

In order to indicate that the formatter is a string that needs to be converted to a JavaScript function, NiceGUI uses the convention of a leading colon ":" for the dictionary key. But pyecharts doesn't know that, so `dump_options_with_quotes` doesn't include such a colon.

Here is a way to solve the problem by changing the key from "formatter" to ":formatter":

```py
options = json.loads(kline.dump_options_with_quotes())
options['tooltip'][':formatter'] = options['tooltip']['formatter']
ui.echart(options)
```

**Comment by TyrionZK** [comment]
Thanks for your quick response. This is very helpful for me as a novice in nicegui, especially when there is relatively less information about nicegui compared to other languages.

According to your instructions，I have modified the function dump_options_with_quotes for JsCode and it works. Now I share it with others. It is here:
```py
def dump_options_with_quotes(chart):
    from pyecharts.charts.base import default
    from pyecharts.commons import utils
    import simplejson as simple_json
    import re
    options_str = simple_json.dumps(chart.get_options(), indent=4, default=default, ignore_nan=True)
    options_str = re.sub("\"(.*?)\": \"--x_x--0_0--(.*?)--x_x--0_0--\"",'\":\g<1>\": \"\g<2>\"',options_str)
    return options_str

chart = Kline(..)
options = json.loads(dump_options_with_quotes(chart))
ui.chart(options)
ui.run()
```
